### PR TITLE
Run tekton operands reconcile in parallel to speed up the whole process

### PR DESCRIPTION
**Pull request by opokornyy**
**What this PR does / why we need it**:

This PR will run tekton operands reconcile in parallel by using goroutines which should speed up the whole process.

**Release note**:
```None

```
Signed-off-by: Ondrej Pokorny <opokorny@redhat.com>

### CODE REVIEW

File: controllers/tektontasks_controller.go [hunk 1]
```diff
@@ -175,17 +176,37 @@ func (r *tektonTasksReconciler) Reconcile(ctx context.Context, req ctrl.Request)
...
-		if err != nil {
-			tektonRequest.Logger.Info(fmt.Sprintf("Operand reconciliation failed: %s", err.Error()))
-			return nil, err
+
+		wg.Add(1)
+		go func(operand operands.Operand, index int) {
+			defer wg.Done()
+			reconcileResults, err := operand.Reconcile(tektonRequest)
+			results[index].err = err
+			results[index].result = append(results[index].result, reconcileResults...)
```
Comment (0xFelix) [inline controllers/tektontasks_controller.go:198]: Why not directly assign it?

`results[index].result, results[index].err = operand.Reconcile(tektonRequest)`

**Comment by ksimon1** [comment]
please can you update commit message why is this change needed, what it improves?

**Comment by 0xFelix** [comment]
What about the dependency problems introduced by parallelizing the deployment of Tasks and Pipelines currently solved by https://github.com/kubevirt/tekton-tasks-operator/pull/83? Did you find another solution for that?

https://github.com/kubevirt/tekton-tasks-operator/pull/60 was merged as a temporary workaround (ugly) until Tekton is able to deploy the `pipeline` ServiceAccount in `^openshift|kube` namespaces. You lifted the restriction of the namespaces in https://github.com/kubevirt/tekton-tasks-operator/pull/82 and this PR is now essentially dependent on a workaround.

I'm not a huge fan of depending on the workaround. Can you solve if differently?

**Comment by ksimon1** [comment]
@0xFelix currently we don't have a way how to do it differently. Until the tekton fixes the deployment of pipeline SA in openshift|kube namespace we have to deploy our own role.

**Comment by 0xFelix** [comment]
@ksimon1 Yes, but when the initial problem is fixed I would have liked to remove the code in `tekton-pipelines`, that creates `ClusterRoles` and `RoleBindings` again.

But even if the pipeline SA is correctly deployed in `openshift|kube` namespaces, the workaround @opokornyy added would still be needed, because the deployment of pipelines is depending on some of the Tasks. (See https://github.com/kubevirt/tekton-tasks-operator/commit/41188db40bced6d7b17346b0620837e838a1fe33 for example). Do we gain something by duplicating roles so we are able to deploy Tasks and Pipelines in parallel? I'd prefer to do it sequentially if there is no easy solution. Users won't probably even notice if Tasks and Pipelines are deployed in parallel or sequentially.

**Comment by ksimon1** [comment]
@0xFelix I agree, that there is no easy solution to make deployment parallel and instead of introducing any new workaround, we can keep it as it is.

**Comment by ksimon1** [comment]
Closing, as we agreed this might introduce some issues

**Pull request closed without merging**